staging/src/k8s.io/apiserver/pkg/features/kube_features.go

```diff
@@ -294,76 +294,156 @@ func init() {
+	ConsistentListFromCache: {
+		{Version: version.MustParse("1.28"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.31"), Default: true, PreRelease: featuregate.Beta},
+	},
 
-	CoordinatedLeaderElection: {Default: false, PreRelease: featuregate.Alpha},
+	CoordinatedLeaderElection: {
+		{Version: version.MustParse("1.31"), Default: false, PreRelease: featuregate.Alpha},
+	},
 
-	EfficientWatchResumption: {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
+	EfficientWatchResumption: {
+		{Version: version.MustParse("1.20"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.21"), Default: true, PreRelease: featuregate.Beta},
+		{Version: version.MustParse("1.24"), Default: true, PreRelease: featuregate.GA, LockToDefault: true},
+	},
 
-	KMSv1: {Default: false, PreRelease: featuregate.Deprecated},
+	MutatingAdmissionPolicy: {
+		{Version: version.MustParse("1.30"), Default: false, PreRelease: featuregate.Alpha},
+	},
 
-	OpenAPIEnums: {Default: true, PreRelease: featuregate.Beta},
+	OpenAPIEnums: {
+		{Version: version.MustParse("1.23"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.24"), Default: true, PreRelease: featuregate.Beta},
+	},
 
-	RemainingItemCount: {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.32
+	RemainingItemCount: {
+		{Version: version.MustParse("1.15"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.16"), Default: true, PreRelease: featuregate.Beta},
+		{Version: version.MustParse("1.29"), Default: true, PreRelease: featuregate.GA, LockToDefault: true},
+	},
 
-	ResilientWatchCacheInitialization: {Default: true, PreRelease: featuregate.Beta},
+	ResilientWatchCacheInitialization: {
+		{Version: version.MustParse("1.31"), Default: true, PreRelease: featuregate.Beta},
+	},
 
-	RetryGenerateName: {Default: true, PreRelease: featuregate.Beta},
+	SeparateCacheWatchRPC: {
+		{Version: version.MustParse("1.30"), Default: true, PreRelease: featuregate.Beta},
+	},
 
-	SeparateCacheWatchRPC: {Default: true, PreRelease: featuregate.Beta},
+	StorageVersionAPI: {
+		{Version: version.MustParse("1.20"), Default: false, PreRelease: featuregate.Alpha},
+	},
 
-	StorageVersionAPI: {Default: false, PreRelease: featuregate.Alpha},
+	StorageVersionHash: {
+		{Version: version.MustParse("1.14"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.15"), Default: true, PreRelease: featuregate.Beta},
+	},
 
-	StorageVersionHash: {Default: true, PreRelease: featuregate.Beta},
+	StrictCostEnforcementForVAP: {
+		{Version: version.MustParse("1.30"), Default: false, PreRelease: featuregate.Alpha},
+	},
 
-	StrictCostEnforcementForVAP: {Default: false, PreRelease: featuregate.Beta},
+	StrictCostEnforcementForWebhooks: {
+		{Version: version.MustParse("1.30"), Default: false, PreRelease: featuregate.Alpha},
```

Comment:
```suggestion
		{Version: version.MustParse("1.30"), Default: false, PreRelease: featuregate.Beta},
```

Reply:
this is a tough PR to eyeball review... is there a way to make sure we're not changing any defaults or prerelease states in the translation?

Reply:
I agree, this PR has been quite challenging (I've made several mistakes due to the large changes involved). I can write a script if you'd like, or perhaps I could split this PR into a few smaller ones to make the review process easier?

Reply:
To ensure that the correct state doesn't drift, we could reuse the update/verify-featuregates.sh to to capture a merged (versioned unioned with unversioned) state and ensure that file does not change when we make these migrations. Historical state will be difficult to verify as we rely on comments that are potentially inaccurate (ref: https://github.com/kubernetes/kubernetes/pull/127200#issuecomment-2335240383) and will need to be manually checked. For these migration PRs, I can ask multiple people to make a pass before merging.

Reply:
```
$ go test ./pkg/features -v -run TestHistory
=== RUN   TestHistory
    history_test.go:16: feature APIResponseCompression not found at 1.7
    history_test.go:16: feature APIServerIdentity at release-1.20 doesn't match:
        	release-1.20: featuregate.FeatureSpec{Default:false, LockToDefault:false, PreRelease:"ALPHA", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"BETA", Version:(*version.Version)(nil)}
    history_test.go:16: feature CPUManager at release-1.26 doesn't match:
        	release-1.26: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:true, PreRelease:"", Version:(*version.Version)(nil)}
    history_test.go:16: feature CustomCPUCFSQuotaPeriod not found at 1.12
    history_test.go:16: feature DRAControlPlaneController not found at 1.26
    history_test.go:16: feature DisableAllocatorDualWrite at release-1.31 doesn't match:
        	release-1.31: featuregate.FeatureSpec{Default:false, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:false, LockToDefault:false, PreRelease:"ALPHA", Version:(*version.Version)(nil)}
    history_test.go:16: feature ElasticIndexedJob at release-1.31 doesn't match:
        	release-1.31: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:true, PreRelease:"", Version:(*version.Version)(nil)}
    history_test.go:16: feature EventedPLEG not found at 1.25
    history_test.go:16: feature KubeProxyDrainingTerminatingNodes at release-1.31 doesn't match:
        	release-1.31: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:true, PreRelease:"", Version:(*version.Version)(nil)}
    history_test.go:16: feature PodSchedulingReadiness at release-1.30 doesn't match:
        	release-1.30: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:true, PreRelease:"", Version:(*version.Version)(nil)}
    history_test.go:16: feature RuntimeClassInImageCriApi not found at 1.29
    history_test.go:16: feature SchedulerQueueingHints not found at 1.27
    history_test.go:16: feature StatefulSetStartOrdinal at release-1.31 doesn't match:
        	release-1.31: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:true, PreRelease:"", Version:(*version.Version)(nil)}
    history_test.go:16: feature TopologyManagerPolicyOptions at release-1.26 doesn't match:
        	release-1.26: featuregate.FeatureSpec{Default:false, LockToDefault:false, PreRelease:"ALPHA", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"BETA", Version:(*version.Version)(nil)}
    history_test.go:16: feature UserNamespacesSupport not found at 1.25
    history_test.go:17: feature APIResponseCompression not found at 1.7
    history_test.go:17: feature APIServerIdentity at release-1.20 doesn't match:
        	release-1.20: featuregate.FeatureSpec{Default:false, LockToDefault:false, PreRelease:"ALPHA", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"BETA", Version:(*version.Version)(nil)}
--- FAIL: TestHistory (0.60s)
```

Reply:
I wrote a unit test we shouldn't merge that scrapes historical state out of git release branches (last commit in https://github.com/liggitt/kubernetes/commits/version_fg_apiserver/)

it's quick and dirty, but several of the things I've spot checked so far were actually wrong and needed fixing (APIResponseCompression gating was added in 1.8, not 1.7, for example)

Reply:
neat, thanks! We'll integrate something like this in to make reviewing easier. 

cc @jpbetz @aaron-prindle: release branches are another good source of truth in addition to the comments and website.

Reply:
pushed an update that checks the full history to check transition points as well, and that features didn't exist prior to the initial release claimed:

```
=== RUN   TestHistory
    history_test.go:19: feature APIServerIdentity at release-1.20 doesn't match:
        	release-1.20: featuregate.FeatureSpec{Default:false, LockToDefault:false, PreRelease:"ALPHA", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"BETA", Version:(*version.Version)(nil)}
    history_test.go:19: unexpected feature APIServerIdentity history diff:
          featuregate.VersionedSpecs{
          	{
        - 		Default:       true,
        + 		Default:       false,
          		LockToDefault: false,
        - 		PreRelease:    "BETA",
        + 		PreRelease:    "ALPHA",
          		Version:       s"1.20",
          	},
        + 	{Default: true, PreRelease: "BETA", Version: s"1.26"},
          }
    history_test.go:19: feature CPUManager at release-1.26 doesn't match:
        	release-1.26: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:true, PreRelease:"", Version:(*version.Version)(nil)}
    history_test.go:19: unexpected feature CPUManager history diff:
          featuregate.VersionedSpecs{
          	{PreRelease: "ALPHA", Version: s"1.8"},
          	{Default: true, PreRelease: "BETA", Version: s"1.10"},
          	{
          		Default:       true,
        - 		LockToDefault: true,
        + 		LockToDefault: false,
          		PreRelease:    "",
          		Version:       s"1.26",
          	},
          }
    history_test.go:19: error building history for CustomCPUCFSQuotaPeriod at release-1.12
    history_test.go:19: feature CustomCPUCFSQuotaPeriod not found at 1.12
    history_test.go:19: unexpected feature CustomCPUCFSQuotaPeriod history diff:
          featuregate.VersionedSpecs{
        - 	{PreRelease: "ALPHA", Version: s"1.12"},
          }
    history_test.go:19: error building history for DRAControlPlaneController at release-1.26
    history_test.go:19: feature DRAControlPlaneController not found at 1.26
    history_test.go:19: unexpected feature DRAControlPlaneController history diff:
          featuregate.VersionedSpecs{
        - 	{PreRelease: "ALPHA", Version: s"1.26"},
          }
    history_test.go:19: feature DisableAllocatorDualWrite at release-1.31 doesn't match:
        	release-1.31: featuregate.FeatureSpec{Default:false, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:false, LockToDefault:false, PreRelease:"ALPHA", Version:(*version.Version)(nil)}
    history_test.go:19: unexpected feature DisableAllocatorDualWrite history diff:
          featuregate.VersionedSpecs{
        - 	{PreRelease: "ALPHA", Version: s"1.31"},
          }
    history_test.go:19: unexpected feature DisableKubeletCloudCredentialProviders history diff:
          featuregate.VersionedSpecs{
          	{PreRelease: "ALPHA", Version: s"1.23"},
          	{Default: true, PreRelease: "BETA", Version: s"1.29"},
        + 	{Default: true, LockToDefault: true, Version: s"1.31"},
          }
    history_test.go:19: feature DynamicResourceAllocation existed prior to first spec entry of 1.29
    history_test.go:19: feature ElasticIndexedJob at release-1.31 doesn't match:
        	release-1.31: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:true, PreRelease:"", Version:(*version.Version)(nil)}
    history_test.go:19: unexpected feature ElasticIndexedJob history diff:
          featuregate.VersionedSpecs{
          	{Default: true, PreRelease: "BETA", Version: s"1.27"},
          	{
          		Default:       true,
        - 		LockToDefault: true,
        + 		LockToDefault: false,
          		PreRelease:    "",
          		Version:       s"1.31",
          	},
          }
    history_test.go:19: error building history for EventedPLEG at release-1.25
    history_test.go:19: feature EventedPLEG not found at 1.25
    history_test.go:19: unexpected feature EventedPLEG history diff:
          featuregate.VersionedSpecs{
        - 	{PreRelease: "ALPHA", Version: s"1.25"},
          }
    history_test.go:19: feature KubeProxyDrainingTerminatingNodes at release-1.31 doesn't match:
        	release-1.31: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:true, PreRelease:"", Version:(*version.Version)(nil)}
    history_test.go:19: unexpected feature KubeProxyDrainingTerminatingNodes history diff:
          featuregate.VersionedSpecs{
          	{PreRelease: "ALPHA", Version: s"1.28"},
          	{Default: true, PreRelease: "BETA", Version: s"1.30"},
          	{
          		Default:       true,
        - 		LockToDefault: true,
        + 		LockToDefault: false,
          		PreRelease:    "",
          		Version:       s"1.31",
          	},
          }
    history_test.go:19: feature PodSchedulingReadiness at release-1.30 doesn't match:
        	release-1.30: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:true, PreRelease:"", Version:(*version.Version)(nil)}
    history_test.go:19: unexpected feature PodSchedulingReadiness history diff:
          featuregate.VersionedSpecs{
          	{PreRelease: "ALPHA", Version: s"1.26"},
          	{Default: true, PreRelease: "BETA", Version: s"1.27"},
          	{
          		Default:       true,
        - 		LockToDefault: true,
        + 		LockToDefault: false,
          		PreRelease:    "",
          		Version:       s"1.30",
          	},
          }
    history_test.go:19: error building history for RuntimeClassInImageCriApi at release-1.29
    history_test.go:19: feature RuntimeClassInImageCriApi not found at 1.29
    history_test.go:19: unexpected feature RuntimeClassInImageCriApi history diff:
          featuregate.VersionedSpecs{
        - 	{PreRelease: "ALPHA", Version: s"1.29"},
          }
    history_test.go:19: error building history for SchedulerQueueingHints at release-1.27
    history_test.go:19: feature SchedulerQueueingHints not found at 1.27
    history_test.go:19: unexpected feature SchedulerQueueingHints history diff:
          featuregate.VersionedSpecs{
        - 	{Default: true, PreRelease: "BETA", Version: s"1.27"},
        - 	{PreRelease: "BETA", Version: s"1.28"},
          }
    history_test.go:19: feature SeparateCacheWatchRPC existed prior to first spec entry of 1.30
    history_test.go:19: feature StatefulSetStartOrdinal at release-1.31 doesn't match:
        	release-1.31: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:true, PreRelease:"", Version:(*version.Version)(nil)}
    history_test.go:19: unexpected feature StatefulSetStartOrdinal history diff:
          featuregate.VersionedSpecs{
          	{PreRelease: "ALPHA", Version: s"1.26"},
          	{Default: true, PreRelease: "BETA", Version: s"1.27"},
          	{
          		Default:       true,
        - 		LockToDefault: true,
        + 		LockToDefault: false,
          		PreRelease:    "",
          		Version:       s"1.31",
          	},
          }
    history_test.go:19: feature TopologyManagerPolicyOptions at release-1.26 doesn't match:
        	release-1.26: featuregate.FeatureSpec{Default:false, LockToDefault:false, PreRelease:"ALPHA", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"BETA", Version:(*version.Version)(nil)}
    history_test.go:19: unexpected feature TopologyManagerPolicyOptions history diff:
          featuregate.VersionedSpecs{
          	{
        - 		Default:       true,
        + 		Default:       false,
          		LockToDefault: false,
        - 		PreRelease:    "BETA",
        + 		PreRelease:    "ALPHA",
          		Version:       s"1.26",
          	},
        + 	{Default: true, PreRelease: "BETA", Version: s"1.28"},
          }
    history_test.go:19: feature UnauthenticatedHTTP2DOSMitigation existed prior to first spec entry of 1.29
    history_test.go:19: error building history for UserNamespacesSupport at release-1.25
    history_test.go:19: feature UserNamespacesSupport not found at 1.25
    history_test.go:19: unexpected feature UserNamespacesSupport history diff:
          featuregate.VersionedSpecs{
        - 	{PreRelease: "ALPHA", Version: s"1.25"},
        - 	{PreRelease: "BETA", Version: s"1.30"},
          }
    history_test.go:19: feature WatchFromStorageWithoutResourceVersion existed prior to first spec entry of 1.30
    history_test.go:20: feature APIServerIdentity at release-1.20 doesn't match:
        	release-1.20: featuregate.FeatureSpec{Default:false, LockToDefault:false, PreRelease:"ALPHA", Version:(*version.Version)(nil)}
        	        HEAD: featuregate.FeatureSpec{Default:true, LockToDefault:false, PreRelease:"BETA", Version:(*version.Version)(nil)}
    history_test.go:20: unexpected feature APIServerIdentity history diff:
          featuregate.VersionedSpecs{
          	{
        - 		Default:       true,
        + 		Default:       false,
          		LockToDefault: false,
        - 		PreRelease:    "BETA",
        + 		PreRelease:    "ALPHA",
          		Version:       s"1.20",
          	},
        + 	{Default: true, PreRelease: "BETA", Version: s"1.26"},
          }
    history_test.go:20: feature SeparateCacheWatchRPC existed prior to first spec entry of 1.30
    history_test.go:20: feature UnauthenticatedHTTP2DOSMitigation existed prior to first spec entry of 1.29
    history_test.go:20: feature WatchFromStorageWithoutResourceVersion existed prior to first spec entry of 1.30
--- FAIL: TestHistory (0.76s)
```

Reply:
> cc @jpbetz @aaron-prindle: release branches are another good source of truth in addition to the comments and website.

yeah, I would say they are the canonical source of truth and we should assume the comments and docs are unreliable

Reply:
pushed another update that handles aliases and renames across branches, and found some issues with earlier migrations of gates to versioned gates that regressed state

the full diff that passes the history test is:
```diff
diff --git a/pkg/features/versioned_kube_features.go b/pkg/features/versioned_kube_features.go
index eb8bb18eac1..0ecae9ac9af 100644
--- a/pkg/features/versioned_kube_features.go
+++ b/pkg/features/versioned_kube_features.go
@@ -125,6 +125,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	DisableKubeletCloudCredentialProviders: {
 		{Version: version.MustParse("1.23"), Default: false, PreRelease: featuregate.Alpha},
 		{Version: version.MustParse("1.29"), Default: true, PreRelease: featuregate.Beta},
+		{Version: version.MustParse("1.31"), Default: true, PreRelease: featuregate.GA, LockToDefault: true},
 	},
 
 	DRAControlPlaneController: {
@@ -132,7 +133,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	DynamicResourceAllocation: {
-		{Version: version.MustParse("1.29"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.26"), Default: false, PreRelease: featuregate.Alpha},
 	},
 
 	ElasticIndexedJob: {
@@ -141,7 +142,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	EventedPLEG: {
-		{Version: version.MustParse("1.25"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.26"), Default: false, PreRelease: featuregate.Alpha},
 	},
 
 	ExecProbeTimeout: {
@@ -172,12 +173,13 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	genericfeatures.APIResponseCompression: {
-		{Version: version.MustParse("1.7"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.8"), Default: false, PreRelease: featuregate.Alpha},
 		{Version: version.MustParse("1.16"), Default: true, PreRelease: featuregate.Beta},
 	},
 
 	genericfeatures.APIServerIdentity: {
-		{Version: version.MustParse("1.20"), Default: true, PreRelease: featuregate.Beta},
+		{Version: version.MustParse("1.20"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.26"), Default: true, PreRelease: featuregate.Beta},
 	},
 
 	genericfeatures.APIServerTracing: {
@@ -232,7 +234,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	genericfeatures.SeparateCacheWatchRPC: {
-		{Version: version.MustParse("1.30"), Default: true, PreRelease: featuregate.Beta},
+		{Version: version.MustParse("1.28"), Default: true, PreRelease: featuregate.Beta},
 	},
 
 	genericfeatures.StorageVersionAPI: {
@@ -263,6 +265,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	genericfeatures.UnauthenticatedHTTP2DOSMitigation: {
+		{Version: version.MustParse("1.25"), Default: false, PreRelease: featuregate.Beta},
 		{Version: version.MustParse("1.29"), Default: true, PreRelease: featuregate.Beta},
 	},
 
@@ -277,7 +280,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	genericfeatures.WatchFromStorageWithoutResourceVersion: {
-		{Version: version.MustParse("1.30"), Default: false, PreRelease: featuregate.Beta},
+		{Version: version.MustParse("1.27"), Default: false, PreRelease: featuregate.Beta},
 	},
 
 	genericfeatures.WatchList: {
@@ -547,7 +550,6 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	SchedulerQueueingHints: {
-		{Version: version.MustParse("1.27"), Default: true, PreRelease: featuregate.Beta},
 		{Version: version.MustParse("1.28"), Default: false, PreRelease: featuregate.Beta},
 	},
 
@@ -635,7 +637,8 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	TopologyManagerPolicyOptions: {
-		{Version: version.MustParse("1.26"), Default: true, PreRelease: featuregate.Beta},
+		{Version: version.MustParse("1.26"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.28"), Default: true, PreRelease: featuregate.Beta},
 	},
 
 	TranslateStreamCloseWebsocketRequests: {
diff --git a/staging/src/k8s.io/apiserver/pkg/features/kube_features.go b/staging/src/k8s.io/apiserver/pkg/features/kube_features.go
index 44dbd74298e..99050ee406f 100644
--- a/staging/src/k8s.io/apiserver/pkg/features/kube_features.go
+++ b/staging/src/k8s.io/apiserver/pkg/features/kube_features.go
@@ -325,12 +325,13 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	APIResponseCompression: {
-		{Version: version.MustParse("1.7"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.8"), Default: false, PreRelease: featuregate.Alpha},
 		{Version: version.MustParse("1.16"), Default: true, PreRelease: featuregate.Beta},
 	},
 
 	APIServerIdentity: {
-		{Version: version.MustParse("1.20"), Default: true, PreRelease: featuregate.Beta},
+		{Version: version.MustParse("1.20"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.26"), Default: true, PreRelease: featuregate.Beta},
 	},
 
 	APIServerTracing: {
@@ -385,7 +386,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	SeparateCacheWatchRPC: {
-		{Version: version.MustParse("1.30"), Default: true, PreRelease: featuregate.Beta},
+		{Version: version.MustParse("1.28"), Default: true, PreRelease: featuregate.Beta},
 	},
 
 	StorageVersionAPI: {
@@ -416,6 +417,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	UnauthenticatedHTTP2DOSMitigation: {
+		{Version: version.MustParse("1.25"), Default: false, PreRelease: featuregate.Beta},
 		{Version: version.MustParse("1.29"), Default: true, PreRelease: featuregate.Beta},
 	},
 
@@ -430,7 +432,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
 	WatchFromStorageWithoutResourceVersion: {
-		{Version: version.MustParse("1.30"), Default: false, PreRelease: featuregate.Beta},
+		{Version: version.MustParse("1.27"), Default: false, PreRelease: featuregate.Beta},
 	},
 
 	WatchList: {
```

Reply:
I believe I’ve addressed everything. Could you point me to where you’ve pushed the new test so I can run it manually?

Reply:
next to last commit in https://github.com/liggitt/kubernetes/commits/version_fg_apiserver/

Reply:
Seems correct:
```
 ~/Desktop/PersonlRepos/kubernetes/ [version_fg_apiserver*] go test ./pkg/features -v -run TestHistory                          
=== RUN   TestHistory
--- PASS: TestHistory (0.98s)
PASS
ok      k8s.io/kubernetes/pkg/features  1.439s
```
Also, git diff shows no changes that need to be addressed.